### PR TITLE
feat: registry credential validation

### PR DESCRIPTION
**Pull request by vzhukovs**
### What does this PR do?
Checks credential before adding them to auth.json. 

Signed-off-by: Vladyslav Zhukovskyi <vzhukovs@redhat.com>

### Screenshot/screencast of this PR

validation

### What issues does this PR fix or reference?

#292 

### How to test this PR?

Try to add a valid registry credential to Add Registry dialog.

### CODE REVIEW

File: extensions/podman/src/registry-setup.ts [hunk 1]
```diff
@@ -196,4 +200,35 @@ export class RegistrySetup {
       });
     });
   }
+
+  protected checkAuthOptions(authOptions: extensionApi.RegistryCreateOptions): Promise<void | Error> {
+    return execPromise(getPodmanCli(), [
```
Comment (benoitf) [inline extensions/podman/src/registry-setup.ts]: would be nice to use REST API
https://docs.docker.com/registry/spec/auth/token/

File: packages/main/src/plugin/image-registry.ts [hunk 2]
```diff
@@ -134,8 +137,8 @@ export class ImageRegistry {
     if (exists) {
       throw new Error(`Registry ${registryCreateOptions.serverUrl} already exists`);
     }
-    const registry = provider.create(registryCreateOptions);
-    this.telemetryService.track('createRegistry', {
+    const registry = await provider.create(registryCreateOptions);
```
Comment (benoitf) [inline packages/main/src/plugin/image-registry.ts]: is that your IDE/tooling that is reporting the issue of missing await ?

File: packages/main/src/plugin/image-registry.ts [hunk 3]
```diff
@@ -134,8 +137,8 @@ export class ImageRegistry {
...
+    await this.telemetryService.track('createRegistry', {
```
Comment (benoitf) [inline packages/main/src/plugin/image-registry.ts]: usually I was not using await to send events as I was not interested in waiting it's actually sent before returning the method

File: extensions/podman/src/registry-setup.ts (hunk 1 above)
Comment (vzhukovs) [reply extensions/podman/src/registry-setup.ts]: updated

File: packages/main/src/plugin/image-registry.ts (hunk 2 above)
Comment (vzhukovs) [reply packages/main/src/plugin/image-registry.ts]: yep, as far as we now having credentials validation in `create` method, we have to wait, when credentials will be checked

File: packages/main/src/plugin/image-registry.ts (hunk 3 above)
Comment (vzhukovs) [reply packages/main/src/plugin/image-registry.ts]: removed await modifier

File: extensions/podman/src/registry-setup.ts [hunk 4]
```diff
@@ -197,3 +204,43 @@ export class RegistrySetup {
     });
   }
 }
+
+function getKnownAuthUrl(serviceUrl: string): string | undefined {
+  // check for includes, cause user may type 'https://quay.io' or 'quay.io'
+  if (serviceUrl.includes('quay.io')) {
+    return 'https://quay.io/v2/auth';
+  } else if (serviceUrl.includes('docker.io')) {
+    return 'https://auth.docker.io/token';
```
Comment (benoitf) [inline extensions/podman/src/registry-setup.ts]: here we need to use www-authenticate header of the registry

Comment (vzhukovs) [reply extensions/podman/src/registry-setup.ts]: done

File: packages/main/src/plugin/image-registry.ts [hunk 5]
```diff
@@ -159,4 +169,71 @@ export class ImageRegistry {
...
+      if (!registryUrl.startsWith('http')) {
+        registryUrl = 'https://' + registryUrl;
+      }
+    }
+
+    try {
+      await got.get(registryUrl);
+    } catch (requestErr) {
+      if (requestErr instanceof HTTPError && requestErr.response?.statusCode === 401) {
+        console.log('Registry ' + registryUrl + ' requires authentication');
```
Comment (benoitf) [inline packages/main/src/plugin/image-registry.ts]: do we need these traces ?

File: packages/main/src/plugin/image-registry.ts [hunk 6]
```diff
@@ -159,4 +169,71 @@ export class ImageRegistry {
...
+        if (authenticate === undefined) {
+          return undefined;
+        }
+
+        const realm = authenticate.match(/realm="(.*?)"/);
+        if (realm === null || realm.length !== 2) {
+          return undefined;
+        }
+
+        console.log('Fetch registry authentication URL: ' + realm[1]);
```
Comment (benoitf) [inline packages/main/src/plugin/image-registry.ts]: and here ?

File: packages/main/src/plugin/image-registry.ts [hunk 7]
```diff
@@ -159,4 +169,71 @@ export class ImageRegistry {
...
     this._onDidUpdateRegistry.fire(Object.freeze(registry));
   }
+
+  async getAuthUrl(serviceUrl: string): Promise<string | undefined> {
+    let registryUrl: string;
+
+    if (serviceUrl.includes('docker.io')) {
+      registryUrl = 'https://index.docker.io/v2/';
+    } else {
+      registryUrl = serviceUrl + '/v2/';
```
Comment (benoitf) [inline packages/main/src/plugin/image-registry.ts]:
```suggestion
      registryUrl = `${serviceUrl}/v2/`;
```

File: packages/main/src/plugin/image-registry.ts [hunk 8]
```diff
@@ -159,4 +169,71 @@ export class ImageRegistry {
...
+
+      if (!registryUrl.startsWith('http')) {
+        registryUrl = 'https://' + registryUrl;
```
Comment (benoitf) [inline packages/main/src/plugin/image-registry.ts]:
```suggestion
        registryUrl = `https://${registryUrl}`;
```

File: packages/main/src/plugin/image-registry.ts (hunk 5 above)
Comment (vzhukovs) [reply packages/main/src/plugin/image-registry.ts]: these traces needs only for the debug purposes, should we keep them or are they redundant?

File: packages/main/src/plugin/image-registry.ts (hunk 7 above)
Comment (vzhukovs) [reply packages/main/src/plugin/image-registry.ts]: updated

File: packages/main/src/plugin/image-registry.ts (hunk 8 above)
Comment (vzhukovs) [reply packages/main/src/plugin/image-registry.ts]: updated

File: package.json [hunk 9]
```diff
@@ -88,7 +88,8 @@
     "moment": "^2.29.2",
     "os-locale": "^6.0.2",
     "stream-json": "^1.7.4",
-    "tar-fs": "^2.1.1"
+    "tar-fs": "^2.1.1",
+    "got": "^12.1.0"
```
Comment (benoitf) [inline package.json]: any reason to not use the latest one ? [12.3.1](https://www.npmjs.com/package/got/v/12.3.1)

File: packages/main/src/plugin/image-registry.ts [hunk 10]
```diff
@@ -159,4 +169,69 @@ export class ImageRegistry {
...
+  async getAuthUrl(serviceUrl: string): Promise<string | undefined> {
+    let registryUrl: string;
+
+    if (serviceUrl.includes('docker.io')) {
+      registryUrl = 'https://index.docker.io/v2/';
+    } else {
+      registryUrl = `${serviceUrl}/v2/`;
+
+      if (!registryUrl.startsWith('http')) {
+        registryUrl = 'https://${registryUrl}';
```
Comment (benoitf) [inline packages/main/src/plugin/image-registry.ts]: looks like it's not using backquotes there

Comment (vzhukovs) [reply packages/main/src/plugin/image-registry.ts]: fixed

File: package.json (hunk 9 above)
Comment (vzhukovs) [reply package.json]: updated

**Review by benoitf — APPROVED** [review]
Tested with quay.io and ghcr.io
image


able to pull images then
image


I noticed that when we have an error displayed at the bottom, the height of the dialog box is changing


https://user-images.githubusercontent.com/436777/187254040-0173bc0c-d181-4810-b0c9-4fd27c5636be.mp4

Would be nice to have a 'skip validation' option

**Pull request merged**